### ISSUE TITLE
Fortran works

### ISSUE THREAD

**Issue by zbeekman** [codecov/support#23]
Just FYI, you could probably lump Fortran in with C and C++ in the documentation. I haven’t tried it yet, but there’s no reason that it won’t work, since any Fortran compiler worth its salt supports gcov. FWIW, right now a [project I contribute to](https://github.com/jacobwilliams/json-fortran) uses coveralls, but as it stands this requires us to use [cpp-coveralls](https://github.com/eddyxu/cpp-coveralls). I’m thinking it might make sense to migrate to codecov.

TL;DR you already support Fortran via C/C++ support so you may as well let people know this.

**Comment by stevepeak** [comment]
Anything I can do to help that transition, please let me know. I look forward to your finding about Fortran. Thank you @zbeekman

**Comment by zbeekman** [comment]
So, it seems that it definitely works: https://codecov.io/github/zbeekman/json-fortran/src/json_module.F90?ref=codecov

The firefox plugin will give the repo coverage percentage and show the line by line coverage :100: :+1: :tada:  https://github.com/zbeekman/json-fortran/blob/codecov/src/json_module.F90

I will be happy to remove the dependency on CPP-coveralls if the project's BDFL approves. Here is my branch where I implemented this:
https://github.com/zbeekman/json-fortran/tree/codecov

As you can see from the latest commit's diff this helped simplified our .travis.yml. Thanks for the cool tool and browser plugins! Now if only we could get sourcegraph to support C/C++/Fortran

If you want to update the documentation to let people know Fortran support works out of the box please feel free to do so. If you prefer I can keep this ticket open until @jacobwilliams decides if it's prudent to migrate to codecov and then you could even link jacobwilliams/json-fortran as an example in the codecov documentation... up to you @stevepeak

**Comment by zbeekman** [comment]
p.s. I'm looking forwards to the safari browser extension!

**Comment by zbeekman** [comment]
@stevepeak : jaconwilliams/json-fortran has now officially migrated to codecov. Feel free to link to the project as an example of supporting Fortran.

**Comment by stevepeak** [comment]
Awesome! Thank you :+1: 

Thank you for the feedback on everything :) I'm so excited you are enjoying this product! The safari extension is challenging, but I hope to get that finished soon. The repo is [open source here](https://github.com/codecov/browser-extension).

I'll add Fortran as a supported language as well. I would like to add an example at [codecov/example-fortran](https://github.com/codecov/example-fortran). Would you be interested in contributing the example? I've invited you to the Codecov org to if you would like :+1: 

Cheers! Will follow up when I get moving on these

**Comment by zbeekman** [comment]
OK, I’ve started making the example-Fortran page… you can check out my progress [here](https://github.com/zbeekman/example-Fortran) Once the DDoS issues get resolved at travis-ci.org and everything is working I’ll push to the codecov repo

**Comment by stevepeak** [comment]
How would I write [this](https://github.com/eddyxu/cpp-coveralls/blob/master/cpp_coveralls/coverage.py#L178-L223) into a bash method. Here is my current template below.

```bash
find . -name '*.gcno' | while read -r i
do
    gcov "$i" &> /dev/null
done
```

> /cc @ionelmc on https://github.com/codecov/codecov-python/issues/28 I'm going to put the `gcov` methods into the bash uploader to start

**Comment by zbeekman** [comment]
as someone who knows bash better than python, I like that you you’re focusing on the bash uploader, FWIW. One issue I have with cpp-coveralls was that it was buggy and/or poorly documented. In particular we had to switch to running gcov manually because even with its exclude options it was running gcov on everything, (read: unit tests) and slurping up all that output and sending it to coveralls. So I think automagically running gcov might be appealing to help reduce the barrier to entry, but it’s important that it’s done correctly and carefully with clear controls for the end user.

Your current example could just use find’s `-exec` flag, fwiw. 

```bash
find . -name ‘*.gcno’ -exec gcov {} \;`
```

Perhaps, the trickier bit is making sure that the user has control over what is and isn’t profiled and can tweak the gcov flags etc…

**Comment by zbeekman** [comment]
@stevepeak I’ve pushed an example to the example-fortran repo. I don’t have access to enable travis for the repo, which will need to happen if you want to show the coverage info on codecov.io for the example hello.f90 program. Hope things look good to you…

**Comment by stevepeak** [comment]
Travis is now enabled on codecov/example-fortran. Thank you :+1: 

----

What if I add an argument to the bash uploader that can ignore directories from being called on.

```sh
bash <(curl -s https://codecov.io/bash) -g skip/this/dir -g skip/this/too
```

```sh
# ...
    "g")
        gcov_ignore="$gcov_ignore
-not -path '$OPTARG'"
        ;;
# ...
bash -c "find . -type f -name '*.gcno' $gcov_ignore" -exec gcov {} \;
# ...
```

View on this commit: https://github.com/codecov/codecov-bash/commit/6154ee720ec967e03533be971fe5531cd7e0c251

----

[Extract from](https://github.com/eddyxu/cpp-coveralls/blob/master/cpp_coveralls/coverage.py#L178-L223)

- What is `--object-directory` used for? Should that be included in Codecov bash?
- Do you see anything else in this script that Codecov bash should need?

**Comment by zbeekman** [comment]
@ionelmc or someone with more C/C++ experience than myself may have some better insight. Fortran is often an after thought for many software tools (although gnu/gcc has really great support)

> What is `—object-directory` used for? Should that be included in Codecov bash?

This appears to just pass along arguments to gcov’s `-o` or `--object-directory` option. I think the use of this is so that when you invoke `gcov` as `gcov … src-file.c` it can locate the `.gcno` files. I think if you just pass it the `.gcno` files directly that the source file location is then embedded in the `.gcno` possibly as a relative path from the directory in which the compiler was invoked. TL;DR: I don’t know for sure if this is necessary. Maybe don’t build it until it’s needed by someone?

> Do you see anything else in this script that Codecov bash should need?

You might need an option to specify an alternater project root. Then you could use bash parameter expansion to default to the CWD.

TBH, I don’t completely understand your rational for the quoting and running a new sub-shell, but my bash-fu may not be as strong as yours.

```bash
find ${proj_root:.} -type f -name '*.gcno' -not -path "${OPTARG}" -exec gcov {} \;
```

I’m not sure I’m the right one to ask here… but in general here are some comments about behavior which may or may not be already implemented in codecov. I know there’s a settings menu on the repository page with an ‘ignore files’ option. For our project (json-fortran) the code base consists of a library and then a number of unit tests. It would be great if coverage information about the unit tests was available when browsing the test programs on github with the codecov browser plugin and on the codecov project page, BUT these tests were _excluded_ from the projects aggregate statistics. (Because in the unit tests theres a lot of code to detect problems etc, that won’t be executed unless a test fails; we care about coverage of the library not executing every line of test code.) This may be the current behavior, but right now I am only running gcov on the library source file/`.gcno` file so as to only produce coverage info for what I care about. If we could exclude the tests from the bulk aggregate then having coverage info for the tests too, could be useful. I hope I’m explaining myself clearly.

**Comment by stevepeak** [comment]
> I don’t completely understand your rational for the quoting and running a new sub-shell

I'm using this method because a user may need to ignore multiple directories which requires separate `-not -path X` methods so I came up with this sub-shell to overcome that challenge.

```bash
# fails
find . -type f -name '*.gcno' -not -path "*/folder/* */folder2/*" -exec gcov {} \;
# works
find . -type f -name '*.gcno' -not -path '*/folder/*' -not -folder '*/folder2/*' -exec gcov {} \;
```

Is there a better way?

> PS I'm just above a nub at bash. Enough to make me dangerous ;-)
